apps/src/blockly/addons/blockSvgLimit.ts

```diff
@@ -0,0 +1,144 @@
+import {BlockSvg} from 'blockly';
+
+const BUBBLE_SIZE = 18;
+const HALF_BUBBLE_SIZE = BUBBLE_SIZE / 2;
+/**
+ * Represents a bubble on a Blockly block, displaying the count of blocks remaining
+ * based on a limit initially stated in the toolbox XML.
+ */
+export default class BlockSvgLimit {
+  protected element_: BlockSvg;
+  protected count: number;
+  protected className: string;
+  protected limitGroup_: SVGElement | undefined;
+  protected limitRect_: SVGElement | undefined;
+  protected limitText_: SVGElement | undefined;
+
+  /**
+   * Constructs an SVG group to track a block limit.
+   * @param {Element} element - The block associated with the limit.
+   * @param {string} [count] - The initial count to display.
+   */
+  constructor(element: BlockSvg, count: number) {
+    this.element_ = element;
+    this.count = count;
+    this.className = 'blocklyLimit';
+
+    this.limitGroup_ = undefined;
+    this.limitRect_ = undefined;
+    this.limitText_ = undefined;
+    this.element_;
+    this.initChildren();
+  }
+
+  /**
+   * Updates the displayed count within the limit bubble and adjusts styling based
+   * on the count.
+   * @param {number} newCount The new count to display.
+   */
+  updateCount(newCount: number) {
+    // Update the count and text content
+    this.count = newCount;
+    this.updateTextAndClass();
+  }
+
+  /**
+   * Updates the text and class of the limit bubble depending on whether the student
+   * has blocks remaining. If over the limit, displays an exclamation mark and changes
+   * the bubble and text color.
+   */
+  updateTextAndClass() {
+    if (!this.limitText_ || !this.limitGroup_) {
+      return;
+    }
+    if (this.count >= 0) {
+      this.limitText_.textContent = `${this.count}`;
+      Blockly.utils.dom.removeClass(this.limitGroup_, 'overLimit');
```

Comment:
Curious: Do we have to use this through the Blockly utils instead of just doing `this.limitGroup_.classList.remove`?

Reply:
No, I don't think we technically need to since they effectively do the same thing. 
https://github.com/google/blockly/blob/blockly-v11.0.0.-beta.6/core/utils/dom.ts#L105-L112

I think I'd lean towards keeping the Blockly util since Blockly is also managing our CSS rules to an extent.